### DISCUSSION TITLE
idea: automatically add repeaters for a location

### DISCUSSION THREAD

**Discussion by DC7IA** [hmatuschek/qdmr#170] [Ideas]
QDMR already uses a database to look up repeaters, so my idea is to allow users to enter a location and then all repeaters within a specific radius (like 50 km) get added to the codeplug, sorting them alphabetically or by distance.

Maybe they could even be added to a specific zone, like a "London zone" or whatever place you're adding. :D 

I think that makes programming radios a lot easier and a lot less work.

Travelling to a place you don't know? Enter a location and all repeaters in that area get added automatically. :D 

What do you think?

**Comment by hmatuschek** [comment]
I thought about that at the very beginning. In earlier versions of qdmr, there was a disabled button in the toolbar labelled "Wizard" to assemble codeplugs in a similar way you described. This however, is not that simple. In particular with the information provided by https://repeatermap.de. The only reliable information I can extract are the frequencies and calls of the repeater. 

For analog channels I also need the CTCSS or DCS settings which are only provided as plain text. Otherwise, it is possible.

For DMR channels it gets more complicated. First of all, I need the color code (again in plaintext in repeatermap). Then I need the network the repeater is associated with. For DMR+ it would be easy, as I would simply create two channels for each repeater with TG9 as the TX contact. For Brandbmeister it is even more complicated: Somehow I need the TX contacts for TS1 and TS2. The latter could be achieved by asking the user for the RX group lists for the two timeslots. Then I could create a channel for each entry of these RX group lists.

I can get this additional information from https://repeaterbook.com. There CTCSS, DCS, colorcode etc is stored in a separate field can could be read reliable. Unfortunately, the access policies of repeaterbook are quiet restrictive. Also, I cannot search repeaterbook for a locator. Just for a country.

**Reply by detournemint** [reply]
Chirp solves this by using a proxy on their own site

**Comment by ra1nb0w** [comment]
maybe a useful site https://www.iz8wnh.it

**Comment by DC7IA** [comment]
Looks like we need a sane and open database... :thinking:

Most databases are terrible. Either terribly outdated or the data isn't normalised which makes processing quite difficult.

Maybe a database with timestamps for updates, so you know data is very old if it hasn't been updated in more than 10 years.

(I recently found a repeater in a database that went offline more than ten years ago...)

Maybe this should be done similar to OpenStreetMap: There should be a simple app allowing the user to select specific things, but not write text. This should ensure CTCSS data isn't a comment.

Definitely out of scope though.

**Comment by ra1nb0w** [comment]
https://www.iz8wnh.it is not enough? you can send your suggestions to  iz8wnh

**Reply by hmatuschek** [reply]
The service appears to be nonfunctional. There are no entries. In the end, repeaterbook might be the most comprehensive option but with the most restrictive data access rules.

**Comment by hmatuschek** [comment]
Similarly to #157, how about moving this issue to discussions?

**Comment by hmatuschek** [comment]
I've marked the #134 as a duplicate of this one, so we have a common topic for the discussion.

**Reply by memtech3** [reply]
Though the issues are related, what I am proposing in #134 is more of a "dumb" workflow that just gives users a better way to manually program a bunch of talkgroups on the same repeater as separate channels in their radios, whereas this discussion seems to be centered around using location data and internet services to aggregate this information for the user.

**Reply by hmatuschek** [reply]
True, but it makes sense to merge this into one step to generate a bunch of channels at once. Concerning the location selection, one could simply resort to use a locator to select all repeaters in a region. My workflow idea would be:

1. Ask the user for the network DMR+ or Brandmeister
2. If Brandmeister was selected
   1. Ask the user for a group list to generate channels for.
   2. Ask the user for a locator. 
   3. Search for all repeater near the locator.
   4. Somehow, we need a way to associate a time-slot to each contact in the selected group list.
   5. Generate a channel for each contact and repeater.
3. If DMR+ was selected
   1. Ask the user for a locator.
   2. Search for all repeater near the locator.
   3. Generate two channels for each repeater (one for each time slot) using TG9 as default TX contact. 

Irrespective of the network. We need some means to infer the color code for each repeater.

**Reply by DC7IA** [reply]
>     1. Ask the user for the network DMR+ or Brandmeister

I'd use checkboxes, that way you just need to do it once for every location and not twice if you want to use both.

**Comment by hmatuschek** [comment]
In general I really like the idea of batch-creating channels for a location. Obviously, there needs to be a separate way to batch-create FM and DMR channels as the workflow is fundamentally different. Lets focus on DMR channels first. 

- [ ] I like the idea to create a channels for a specified location. The simplest way would be to let the user enter a locator. However, just entering "London" would be cool. Wikipedia provides some machine readable metadata for that (see https://www.wikidata.org). 
- [ ] Once the geo information is present, a list of repeaters might be presented that are closest to the specified location.
- [ ] As suggested by @memtech3, the next step would be some means to specify the talk groups. Then qdmr would create a channel for each repeater/talkgroup combination. Here, however, we have an issue. In Brandmeister networks, TS1 is for "long haul" calls and TS2 for regional ones. Qdmr does not know which talk group is regional to the selected location or not. So there must be some means for the user to specifiy that. 
- [ ] Finally, what is about DMR+? There the batch-creation is much easier. TX contact is always TG9. So maybe the first step in the wizard could be the selection "FM", "DMR+" and "Brandmeister DMR"

**Reply by DC7IA** [reply]
>     * Qdmr does not know which talk group is regional to the selected location or not. So there must be some means for the user to specifiy that.

Or take that from the Brandmeister database

**Comment by hmatuschek** [comment]
An alternative source for geo-information might also be the OpenStreetMap project and their API https://wiki.openstreetmap.org/wiki/Overpass_API/Overpass_QL. 

In fact, sending a post request to https://overpass-api.de/api/interpreter with
```
[out:json];
node["name"~"^London"]["place"~"^(city|town)$"];
out body;
```
Gets 6 results including the geo location and some additional meta information to identify these places.
